### ISSUE TITLE
Methods to troubleshoot UnobservedPromiseException

### ISSUE THREAD

**Issue by drew-512** [timcassell/ProtoPromise#283]
Hi Tim and friends,

I'm getting a `UnobservedPromiseException` show up a minute or two after I've stopped Unity when GC catches up.  This means I have zero info on where or why my code is not resolving a deferred promise.

Is there any way to have `PROMISE_DEBUG` record the top of stack trace when a deferred promise is created?  Another idea would be to allow a debug label (string) to be attached to a  promise.  In non-debug mode, the string is just dropped and not part of a promise struct, but in debug mode it would travel around with the promise allowing the developer to see its origin.

Drew

**Comment by timcassell** [comment]
Hi Drew,

`UnobservedPromiseException` means a promise was not awaited or forgotten. If a deferred was left pending, it would instead be a `UnhandledDeferredException`.

The exception message should explain how to enable causality traces to find the culprit. You need to be in DEBUG mode and set `Promise.Config.DebugCausalityTracer = TraceLevel.All`.

**Comment by drew-512** [comment]
Ah, ok thanks for the correction there.

I did have `Promise.TraceLevel.All` set, however it's unclear what I need to look at for help.  There are a ton of fields and the ones that are interesting like the stack trace are truncated by VSCode.   

What and where should I look for more information?

![Screen Shot 2023-10-18 at 2 28 38 PM](https://github.com/timcassell/ProtoPromise/assets/35852900/065f7895-098a-4dba-ab88-f06cf8e2d852)

**Comment by timcassell** [comment]
Oh, it looks you even enabled developer mode! That shouldn't be necessary for you, that's really for me to debug the internal promise code.

The information you're looking for is in the Stacktrace of the `UnhandledException`, in this case the one that contains the `UnobservedPromiseException` as its `InnerException`.

**Comment by timcassell** [comment]
I think VS has options to view the entire stacktrace if you right-click it. Otherwise, you could try grabbing the `ToString()` of the entire exception, which should include all of the stacktrace information.

**Comment by timcassell** [comment]
Also, when it's printed to the console in Unity, it should include the full stacktrace there, also. If it's truncated there, you can view the full thing in the log file (the console has a button that you can click to open it directly).

**Comment by drew-512** [comment]
I wasn't getting any info from Unity since the exception was happening when play mode was starting (when GC was likely happening from the last play).   This means any exception msgs pushed to Unity were being cleared as play mode begins.   Meanwhile, VSCode / C# is still highly unimpressive with strings in the debugger.  I ended up just putting in some lines that print the stack trace to a Debug.Log message.  Given this somewhat likely scenario (where GC occurring right before entering play mode), perhaps an enhancement here makes sense.

Anyway, the offender suprisingly is `await Promise.SwitchToForeground()`, made from an `async void` proc that exits during app shutdown (play mode exit):

```.cs
async void sessionReader(NetworkStream stream) {
    await Promise.SwitchToBackground();
    string logMsg = null;
    try {
        var hdr = new byte[TxMsg.TxHeaderSz];
        int bytesRead;                     
        
        while ((bytesRead = stream.Read(hdr, 0, TxMsg.TxHeaderSz)) == TxMsg.TxHeaderSz) {
            var tx = TxMsg.Acquire(hdr);
            if ((bytesRead = stream.Read(tx.TxStore, TxMsg.TxBodyOfs, tx.TxBodyLen)) != tx.TxBodyLen) {
                throw new Exception($"Failed to read TxMsg body: {bytesRead} != {tx.TxBodyLen}");
            }
            handleIncomingTx(tx);
        }
    } catch (Exception ex) {
        logMsg = ex.ToString();
    }

    // Cancel the writer if the reader stops
    Cancel();
    
    await Promise.SwitchToForeground();
    if (logMsg != null) {
        Logger.Warn(logMsg);
    }
    
    // Tell controller we're fully disconnected
    _client.OnConnectionClosed();
}
```

**Comment by timcassell** [comment]
> I wasn't getting any info from Unity since the exception was happening when play mode was starting (when GC was likely happening from the last play). This means any exception msgs pushed to Unity were being cleared as play mode begins.

You can fix that by disabling clear on play mode.

<img width="138" alt="Screen Shot 2023-10-18 at 4 35 36 PM" src="https://github.com/timcassell/ProtoPromise/assets/35501420/e9cb864a-6981-4b96-af86-53985e222ba2">

> Anyway, the offender suprisingly is await Promise.SwitchToForeground(), made from an async void proc that exits during app shutdown (play mode exit):

I'm guessing that's because you have AppDomain reload disabled. I have a check in the default exception handler to ignore any further exceptions after `OnApplicationQuit`. Since it occurs when you start the game again, it no longer considers it to be in the quitting state, so it does log it.

I can't think of a nice way to handle that case, I'm open to suggestions.

**Comment by drew-512** [comment]
Hmm, is there any way to catch or prevent whatever is going on with that `await Promise.SwitchToForeground()`?

**Comment by timcassell** [comment]
> Hmm, is there any way to catch or prevent whatever is going on with that `await Promise.SwitchToForeground()`?

Suffice to say, there's nothing wrong with your code, and you won't see this issue in production. This is purely a side effect of stopping and restarting code without resetting its run context (AppDomain reload here, process isolation in built player).

**Comment by timcassell** [comment]
Perhaps I could add an API like `Promise.Manager.ResetRuntimeContext()` that can be called from `OnApplicationQuit`. That would then ignore exceptions from every object created before `ResetRuntimeContext` was called.

**Comment by drew-512** [comment]
If it's interesting, Unity seems to be being mean to your exceptions.
![Screen Shot 2023-10-18 at 5 51 17 PM](https://github.com/timcassell/ProtoPromise/assets/35852900/73f0ea66-733d-44e2-acf0-8f10fe36839a)

**Comment by timcassell** [comment]
That looks like it's from throwing the exception on a ThreadPool thread after the default exception handler is removed from `OnApplicationQuit`. Kinda weird that they truncate the inner exceptions, but oh well.

**Comment by timcassell** [comment]
So I tried to repro that code in a test case, and it passes every time I run it. I debugged it and saw the promise was marked as awaited (as it should be). The only thing I can think of is the background thread is being aborted in the middle of the await. I'm unaware of Unity aborting background threads without the AppDomain being reloaded, are you perhaps running your own background SynchronizationContext and aborting threads yourself?

```cs
private class StrongBox<T>
{
    public T value;
}

[MethodImpl(MethodImplOptions.NoInlining)]
private static async void RunAndSwitchContext(StrongBox<bool> box, SynchronizationContext context)
{
    await Promise.SwitchToBackground();
    box.value = true;
    await Promise.SwitchToContext(context);
    box.value = true;
}

[Test]
public void PromiseSwitchToContext_NeverExecuted_NoExceptions(
    [Values] bool forceAsync)
{
    // We're purposefully not completing the promise, to simulate Unity's enter/exit playmode.
    TestHelper.DisableTrackObjectsForRelease();

    var box = new StrongBox<bool>();
    RunAndSwitchContext(box, new NeverExecutedSynchronizationContext());
    while (!box.value) { Thread.MemoryBarrier(); }
    box.value = false;
    Thread.Sleep(1000);
    Assert.False(box.value);

    TestHelper.GcCollectAndWaitForFinalizers();
}
```

**Comment by timcassell** [comment]
@drew-512 Can you confirm if aborting threads is the cause?

**Comment by timcassell** [comment]
So you're not setting anything in `Promise.Config`? Just using the default values? I'm trying to determine the root cause of the issue, and if it's not thread aborts, I can't think of what it might be.

**Comment by drew-512** [comment]
If I'm hearing you right, you'd like a repro project of the shows a warning as in my Oct 18 post?

I have to sheepishly admit that I didn't event know about `Promise.Config` until your mention of it above (and so had not made any changes to that).

**Comment by timcassell** [comment]
> If I'm hearing you right, you'd like a repro project of the shows a warning as in my Oct 18 post?

If possible, yes. I'm uncomfortable merging #303 without understanding the actual cause.

> I have to sheepishly admit that I didn't event know about `Promise.Config` until your mention of it above (and so had not made any changes to that).

No worries, that's good to know.

**Comment by drew-512** [comment]
Sorry for the delay, been a tough go personally lately.

Ok, I'll use the latest release version (not branch)?  What vars should I set exactly out of `Promise.Config` to give you the output or stack that you're looking for?

**Comment by timcassell** [comment]
You didn't change any defaults before so no need to do so now. I just need to be able to repro it like you had before, then I can start debugging it myself.

**Comment by timcassell** [comment]
Nevermind, I was able to reproduce it with this code in editor.

```cs
void Start()
{
    for (int i = 0; i < Environment.ProcessorCount * 100; ++i)
    {
        Func().Forget();
    }
}

private async Promise Func()
{
    while (true)
    {
        await Promise.SwitchToBackground();
        Thread.Sleep(10);
        await Promise.SwitchToForeground();
    }
}
```

**Comment by timcassell** [comment]
Aha, looks like it was fixed in #306! To confirm, I added a try/catch:

```cs
try
{
    await Promise.SwitchToForeground();
}
catch (Exception e)
{
    Debug.LogException(e);
}
```

And that actually logs
`InvalidOperationException: SynchronizationOption.Foreground was provided, but Promise.Config.ForegroundContext was null. You should set Promise.Config.ForegroundContext at the start of your application (which may be as simple as 'Promise.Config.ForegroundContext = SynchronizationContext.Current;').`

Damn Unity sometimes logging background thread exceptions, and sometimes not!

@drew-512 If you try latest master, it should be fixed.

**Comment by drew-512** [comment]
Yay, nice work!   Switching to main!
